### PR TITLE
[CARBONDATA-3852] [CARBONDATA-3851] Fix Data update issue with SCD/CDC  for partition table On spark cluster deploy mode

### PR DESCRIPTION
**Pull request by Indhumathi27**
### Why is this PR needed?

Issue 1:
On process IUD with mergeDataSetCommand, we are setting **direct write to store path** carbon property as **true** in executor side and resetting the value to **false** on driver side. Because of this, Update operation is unsuccessfull, as **Issue 2** will come into picture.

Issue 2:
Currently, For insert/load data into partition table, data file will be moved to partition path and index files will remain in tmp file to improve commit job performance. 
When **direct write to store path** is enabled, data files are not  moved to partition path and  load metrics are not set and added to task context. On commitJob, it just return, as index and carbon size will be 0. Thus, the corresponding load is not successful.
 
 ### What changes were proposed in this PR?

For Issue 1:
Set  **direct write to store path** carbon property to default value on close writer while processIUD for merge data set command

For Issue 2:
1. While writing data file, check if partition table and dir path is "partition tmp path", then write data file directly to partiiton path.
2. On close writer, add partition carbon data and index info to metrics
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - NO

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3354/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1613/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1615/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3356/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3357/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1616/

**Comment by akashrn5** [comment]
@Indhumathi27 
1. I think, this is not specific to cdc, i think problem can come in case of normal partition table insert into flow
also when enableDirectlyWriteDataToStorePath is true for writer. Please check that.

2. In case of cdc , we set direct write to hdfs as false before doing insert, please check who made it enabled.

if first point is true, you can change the PR description and remove cdc from there.

**Comment by Indhumathi27** [comment]
@akashrn5 Fixed and updated the PR. PLease review and merge

File: processing/src/main/java/org/apache/carbondata/processing/store/writer/AbstractFactDataWriter.java [hunk 1]
```diff
@@ -318,6 +335,13 @@ public void initializeWriter() throws CarbonDataWriterException {
     try {
       FileFactory.mkdirs(model.getCarbonDataDirectoryPath());
       if (enableDirectlyWriteDataToStorePath) {
+        if (model.getTableSpec().getCarbonTable().isHivePartitionTable() && model
+            .getCarbonDataDirectoryPath().endsWith("tmp")) {
+          // set carbonData file store path to partition path instead of tmp directory
+          carbonDataFileStorePath = model.getCarbonDataDirectoryPath()
+              .substring(0, model.getCarbonDataDirectoryPath().lastIndexOf("/")) + File.separator
```
Comment (akashrn5) [inline processing/src/main/java/org/apache/carbondata/processing/store/writer/AbstractFactDataWriter.java:342]: use `CarbonCommonConstans.FILE_SEPARATOR`

File: processing/src/main/java/org/apache/carbondata/processing/store/writer/AbstractFactDataWriter.java [hunk 2]
```diff
@@ -318,6 +335,13 @@ public void initializeWriter() throws CarbonDataWriterException {
     try {
       FileFactory.mkdirs(model.getCarbonDataDirectoryPath());
       if (enableDirectlyWriteDataToStorePath) {
+        if (model.getTableSpec().getCarbonTable().isHivePartitionTable() && model
+            .getCarbonDataDirectoryPath().endsWith("tmp")) {
```
Comment (akashrn5) [inline processing/src/main/java/org/apache/carbondata/processing/store/writer/AbstractFactDataWriter.java:339]: same as above,` .tmp`

File: processing/src/main/java/org/apache/carbondata/processing/store/writer/AbstractFactDataWriter.java [hunk 3]
```diff
@@ -440,9 +464,23 @@ protected void writeIndexFile() throws IOException, CarbonDataWriterException {
           .copyCarbonDataFileToCarbonStorePath(indexFileName, model.getCarbonDataDirectoryPath(),
               fileSizeInBytes, metrics);
       FileFactory.deleteFile(indexFileName);
+    } else if (model.getTableSpec().getCarbonTable().isHivePartitionTable() && model
+        .getCarbonDataDirectoryPath().endsWith("tmp")) {
```
Comment (akashrn5) [inline processing/src/main/java/org/apache/carbondata/processing/store/writer/AbstractFactDataWriter.java:468]: same as above

**Comment by akashrn5** [comment]
@Indhumathi27 please update the PR description, no  need to cdc, can say partition issue when direct write to hdfs enabled, and in detailed description part can list issues.

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala [hunk 4]
```diff
@@ -283,6 +284,13 @@ case class CarbonMergeDataSetCommand(
           val queue = new util.LinkedList[InternalRow]()
           override def hasNext: Boolean = if (!queue.isEmpty || iter.hasNext) true else {
             writer.close()
+            // revert load direct write to store path after IUD
```
[the remark this answers is not in the document]
Comment (Indhumathi27) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala]: done

File: processing/src/main/java/org/apache/carbondata/processing/store/writer/AbstractFactDataWriter.java (hunk 1 above)
Comment (Indhumathi27) [reply processing/src/main/java/org/apache/carbondata/processing/store/writer/AbstractFactDataWriter.java]: done

File: processing/src/main/java/org/apache/carbondata/processing/store/writer/AbstractFactDataWriter.java (hunk 2 above)
Comment (Indhumathi27) [reply processing/src/main/java/org/apache/carbondata/processing/store/writer/AbstractFactDataWriter.java]: done

File: processing/src/main/java/org/apache/carbondata/processing/store/writer/AbstractFactDataWriter.java (hunk 3 above)
Comment (Indhumathi27) [reply processing/src/main/java/org/apache/carbondata/processing/store/writer/AbstractFactDataWriter.java]: done

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1630/

**Comment by Indhumathi27** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3372/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3377/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1636/

**Comment by Indhumathi27** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3380/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1639/

**Comment by akashrn5** [comment]
LGTM

**Pull request closed without merging**